[REVIEW] Bugs in dbscan concerning memory and results

MatthiasKohl opened the pull request:
This pull request addresses a few issues with the current DBScan implementation:
1. There was a bug concerning the results (issue #2206). This is fixed but it requires slightly more memory: the core-points boolean array is not over the batch only anymore, instead it needs to be as large as the number of input rows. This shouldn't be a big concern, I bench-marked the new implementation and the difference is not noticeable.
EDIT: forgot to mention here that the `Lambda filter_op` in `weak_cc` kernels now works over global indexes, rather than the local batch-wise indexes.
2. I addressed issue #2164 
3. About issue #2163: I improved the estimation of memory (see point 4 below). I also investigated the indexing used in DBScan: the thing is that the output datatype (type of the labels) doesn't need to be used as a way of addressing the pointers/arrays. The output datatype only has to be as large as the number of points, but addresses used for the dense adjacency matrix as well as the sparse matrix (the exclusive scan and the total sum of neighbors) can be larger than the number of points. In the end, it's not worth it to change the implementation here, because very large batches actually don't seem beneficial (on a V100 there is a point where using more memory actually causes a small performance drop). This needs further investigation and I'll open another issue if necessary.
4. I added a parameter allowing to specify how many neighbors one expects for the batch count. This can make the batch size prediction much tighter given the allowed max memory.
5. Some minor refactoring: The typename `Type` was unnecessary and only used for `minPts`, where `Index_` can be used instead. The way that batch count was calculated was a bit weird. I simplified the implementation.

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

MatthiasKohl commented:
rerun tests

cjnolet commented on python/cuml/test/test_dbscan.py:
```diff
@@ -137,3 +137,104 @@ def test_dbscan_out_dtype_fails_invalid_input():
+
+
+def test_core_point_prop1():
+    params = {'eps': 1.1, 'min_samples': 4}
```
Were these tests able to reproduce the core points bug here? They seem pretty simple and it would be nice to see a simple gtest that could catch this issue in the future.

MatthiasKohl replied:
Yes exactly, you can run these tests on the current implementation (without this PR) and they should fail. These tests are specifically designed to fail the current implementation.
And I agree with you, it was a bit faster to code these in Python for me, as I can see the data in space in PyCharm, so that's why I put them there, but they should ultimately go into `cpp/test/sg/dbscan_test.cu`, right?

teju85 replied:
yes, these should end in dbscan_test.cu

teju85 commented on cpp/src/dbscan/runner.cuh:
```diff
@@ -173,9 +173,9 @@ size_t run(const ML::cumlHandle_impl& handle, Type_f* x, Index_ N, Index_ D,
       adj_graph.resize(adjlen, stream);
```
If we are using `curradjlen` below, then this `resize` method should also be using `curradjlen`, right?

MatthiasKohl replied:
This was actually a bit confusing for me as well. Essentially, `adjlen` is actually the maximal sparse adjacency matrix length (so maybe we should rename it `maxAdjlen`?). And it does make sense not to `resize` in later batches with smaller size as this saves some memory ops, as the adj-graph alogs properly over-write this space, even if it's too large.
The only inconvenience right now is that if there is a bug in adj-graph (or in the weak_cc kernel processing it) then cuda-memcheck cannot always find it.

teju85 replied:
>  The only inconvenience right now is that if there is a bug in adj-graph (or in the weak_cc kernel processing it) then cuda-memcheck cannot always find it.

Ideally, I'd expect running cuda-memcheck on unit-tests for dbscan to be able to catch these.

teju85 replied:
Agreed. I think it is better to rename this variable to `maxadjlen`.

MatthiasKohl replied:
I have changed this.

teju85 commented on cpp/src/dbscan/dbscan.cuh:
```diff
@@ -21,45 +21,74 @@
+  // => nBatches >= n_rows * n_rows / MAX_LABEL
+  Index_ nBatchesPrec =
+    (Index_)ceildiv<size_t>((size_t)n_rows * n_rows, MAX_LABEL);
+  if (nBatchesPrec >= 4 * nBatches) {
```
Why the x4 factor? The warning message below doesn't seem to mention this.

MatthiasKohl replied:
Thank you for catching this, it's arbitrary indeed and I should have removed this, since the final benchmarks actually don't really show a benefit of using very large batches with 64-bit labeling.
In general, the idea here was that if the batch-size is limited only by the indexing but there is enough memory for much larger batches available, we should warn the user as for smaller batch sizes, there is actually a serious performance drop when using "too small" batches.
However, in the final benchmarks seem to indicate that when we're actually limited by the indexing, larger batch sizes don't seem to matter anymore, at least right now. So I should probably just move this to a comment until further tuning and bench-marking is done.

teju85 commented on cpp/src/dbscan/dbscan.cuh:
```diff
@@ -21,45 +21,74 @@
+  size_t MAX_LABEL = (size_t)std::numeric_limits<Index_>::max();
+  // n_rows * n_rows_per_batch < MAX_LABEL => n_rows * (n_rows / nBatches) < MAX_LABEL
+  // => nBatches >= n_rows * n_rows / MAX_LABEL
+  Index_ nBatchesPrec =
```
Not sure I got the reasoning behind using `MAX_LABEL` for computing this batch count. Can you elaborate?

MatthiasKohl replied:
This is still actually the same as before. Line 52 (for now) of `dbscan.cuh` has this exact test:
`batchSize * n_rows < MAX_LABEL`
And line 107 (for now) of `runner.cuh` also tests for this : `N * batchSize < MAX_LABEL`

For now, I haven't changed the indexing/addressing used in the actual algorithms because there doesn't seem to be a benefit after all (see also the comment above). So we still have to limit the batch-size according to `MAX_LABEL` even if we have more than 2GB of memory available and we estimate to use more than 2GB per batch.

teju85 replied:
makes sense now. Thanks.

MatthiasKohl commented:
could you review the latest changes again?
@cjnolet I moved the tests to test_dbscan.cu as requested.
@teju85 I moved the warning to a comment and renamed the `adjlen` parameter

MatthiasKohl commented:
Thanks Corey, there are some conflicts now so I'll have to fix that anyway.
Let me just put the Python tests back in, they were passing before.
